Hotwebmatter/add gulp custom command

hotwebmatter opened the pull request:
Build Drupal theme assets with Gulp -- with minor modifications, this approach will work for other frameworks (WordPress, etc.) and other front-end build tools.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/drud/ddev-contrib?pullRequest=123) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/drud/ddev-contrib?pullRequest=123) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/drud/ddev-contrib?pullRequest=123) it.</sub>

albertvolkman commented on custom-commands/gulp:
```diff
@@ -0,0 +1,20 @@
+## Example: "ddev gulp", "ddev gulp styles", "ddev gulp watch"
+
+## Set custom theme root (example works for Drupal)
+DDEV_THEMEROOT=/var/www/html/web/themes/custom/customtheme
```
Maybe something like this to simplify the variable, and not assume the file structure.

```
THEME=customtheme

pushd ${DDEV_DOCROOT}/{$THEME}
```

hotwebmatter replied:
@albertvolkman Thanks for the advice! After talking to Mike Anello about this, I think `${DDEV_DOCROOT}` in a Drupal recommended project is just `web`. So maybe I'll do something like this:

```bash
## Set relative path to custom theme root (example works for Drupal)
THEME=/themes/custom/customtheme

pushd /var/www/html/${DDEV_DOCROOT}/${THEME}
```

This would allow users to configure their docroot to be something other than `web`.

I'll need to test this before I modify the script -- as it is, it works for me!

hotwebmatter replied:
I tried this:

```bash
THEME=themes/custom/customtheme
DDEV_THEMEROOT=/var/www/html/${DDEV_DOCROOT}/${THEME}

pushd ${DDEV_THEMEROOT}
```

But it fails with the following error:

```
 ddev gulp && ddev drush cr            
/mnt/ddev_config/commands/web/gulp: line 10: pushd: /var/www/html//themes/custom/customtheme: No such file or directory
/mnt/ddev_config/commands/web/gulp: line 11: ./node_modules/gulp/bin/gulp.js: No such file or directory
/mnt/ddev_config/commands/web/gulp: line 12: popd: directory stack empty
Failed to run gulp : exit status 1
```

Apparently, ${DDEV_DOCROOT} is not available in the scope of the `.ddev/commands/web/gulp` command script.

rfay replied:
It looks to me like $DDEV_DOCROOT has been there for quite a long time. What ddev version are you running?

hotwebmatter replied:
Hi @rfay ! Thanks for jumping in to help troubleshoot this.

I agree that the documentation indicates this environment variable should be available:
https://ddev.readthedocs.io/en/stable/users/extend/custom-commands/#environment-variables-provided

Just going by the output above, it looks to me that the command `pushd ${DDEV_THEMEROOT}` is attempting to push the directory path `/var/www/html//themes/custom/customtheme`.

That would mean that the variable assignment statement `DDEV_THEMEROOT=/var/www/html/${DDEV_DOCROOT}/${THEME}` is returning the value for `${THEME}` correctly, but returning nothing for the value of `${DDEV_DOCROOT}`.

My custom command works fine when I hard-code the entire path.

To answer your direct question:

```
$ ddev --version
ddev version v1.14.2
```

I believe there is a new version available, which I have not yet installed.

Let me know if there's anything else I can do to troubleshoot the environment variable substitution.

Thanks for all the work that you do to make DDEV such a useful tool!

rfay replied:
Yes, the current stable version of ddev is v1.16.7. Please test with the current stable version.

hotwebmatter replied:
OK, I have updated to the latest stable DDEV-Local:

```
$ ddev version
DDEV-Local version	v1.15.3                                
commit            	v1.15.3                                
db                	drud/ddev-dbserver-mariadb-10.2:v1.15.1
dba               	phpmyadmin/phpmyadmin:5                
ddev-ssh-agent    	drud/ddev-ssh-agent:v1.15.1            
docker            	20.10.5                                
docker-compose    	1.25.0-rc4                             
os                	linux                                  
router            	drud/ddev-router:v1.15.1               
web               	drud/ddev-webserver:v1.15.3             
```

I have added some `echo` statements for debugging:
```bash
#!/bin/bash

## Description: Run gulp in theme directory.
## Usage: gulp styles|watch
## Example: "ddev gulp", "ddev gulp styles", "ddev gulp watch"

THEME=themes/custom/customtheme
DDEV_THEMEROOT=/var/www/html/${DDEV_DOCROOT}/${THEME}

echo "DDEV_DOCROOT: $DDEV_DOCROOT"
echo "DDEV_DOCROOT: ${DDEV_DOCROOT}"
echo "THEME: ${THEME}"
echo "DDEV_THEMEROOT: ${DDEV_THEMEROOT}"

pushd ${DDEV_THEMEROOT}
./node_modules/gulp/bin/gulp.js $1
popd
exit
```

I tried it with and without the curly braces -- not sure what else to try.

When I run this, I see the following output:
```
$ ddev gulp                  
DDEV_DOCROOT: 
DDEV_DOCROOT: 
THEME: themes/custom/customtheme
DDEV_THEMEROOT: /var/www/html//themes/custom/customtheme
/mnt/ddev_config/commands/web/gulp: line 15: pushd: /var/www/html//themes/custom/customtheme: No such file or directory
/mnt/ddev_config/commands/web/gulp: line 16: ./node_modules/gulp/bin/gulp.js: No such file or directory
/mnt/ddev_config/commands/web/gulp: line 17: popd: directory stack empty
Failed to run gulp : exit status 1
```

I'll try adding the `echo ${DDEV_DOCROOT}` to another command to see if the global works elsewhere.

rfay replied:
The latest stable is v1.16.7.

hotwebmatter replied:
I created `.ddev/commands/web/printvars` for more debugging info.

Some environment variables work, others do not:

```bash
#!/bin/bash

## Description: Print DDEV environment variables.
## Usage: printvars
## Example: "ddev printvars"

echo "DDEV_HOST_DB_PORT: ${DDEV_HOST_DB_PORT}"
echo "DDEV_HOST_WEBSERVER_PORT: ${DDEV_HOST_WEBSERVER_PORT}"
echo "DDEV_HOST_HTTPS_PORT: ${DDEV_HOST_HTTPS_PORT}"
echo "DDEV_DOCROOT: ${DDEV_DOCROOT}"
echo "DDEV_HOSTNAME: ${DDEV_HOSTNAME}"
echo "DDEV_PHP_VERSION: ${DDEV_PHP_VERSION}"
echo "DDEV_WEBSERVER_TYPE: ${DDEV_WEBSERVER_TYPE}"
echo "DDEV_PROJECT_TYPE: ${DDEV_PROJECT_TYPE}"
echo "DDEV_ROUTER_HTTP_PORT: ${DDEV_ROUTER_HTTP_PORT}"
echo "DDEV_ROUTER_HTTPS_PORT: ${DDEV_ROUTER_HTTPS_PORT}"

exit
```

When I execute this, I see the following output:
```
$ ddev printvars
DDEV_HOST_DB_PORT: 
DDEV_HOST_WEBSERVER_PORT: 
DDEV_HOST_HTTPS_PORT: 
DDEV_DOCROOT: 
DDEV_HOSTNAME: 
DDEV_PHP_VERSION: 7.3
DDEV_WEBSERVER_TYPE: nginx-fpm
DDEV_PROJECT_TYPE: drupal8
DDEV_ROUTER_HTTP_PORT: 80
DDEV_ROUTER_HTTPS_PORT: 443
```
:thinking:

rfay replied:
That's exactly the right technique. But please use the current stable v1.16.7. If you need help upgrading, please ask. If you're using homebrew, please read the complaint it may be giving you about shallow clone when you `brew update`.

You can also test with a simple `ddev exec 'echo $DDEV_DOCROOT'`

But it would be a lot easier in your custom command to do `set | grep DDEV`

hotwebmatter replied:
Thanks, this is resolved -- it was not the shallow clone problem, I think I just forgot to `ddev poweroff` when upgrading.

The environment variables work now! :tada:

hotwebmatter commented:
I refactored the variables to account for project docroot, per Albert's suggestion.

I also added some sanity checks to prevent changing to a non-existent directory or executing a non-existent `gulp.js`.

hotwebmatter commented:
@rfay Thanks for writing the `README.md` for me! :+1:

rfay commented:
Thanks for writing this whole thing up!

hotwebmatter commented:
Well then, thanks for DDEV-Local!
